iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp

```diff
@@ -203,94 +184,267 @@ static bool isClonableIntoDispatchOp(Operation *op) {
+        resultDynamicDimsPos += numDynamicDims;
+      }
+      resultPos++;
     }
-    rewriter.create<IREE::Flow::ReturnOp>(loc);
   }
-
   LLVM_DEBUG(llvm::dbgs() << "Created dispatchOp shell \n"
                           << *dispatchOp << "\n");
-  return {dispatchOp, clonedOp};
+  return clonedOps;
 }
 
-// Fuses producers marked in the same group recursively.
-//
-// The impl does not worry about the dispatchOp, operands and arguments are set
-// in a post-pattern `legalizeDispatchWorkgroupOperands` function.
-// To simplify the implementation of the dispatch region formation, we just
-// clone the op that needs to be fused inside the dispatch region and just fuse
-// that one. This avoid any concerns related to tensor operands that are only
-// used for their DimOp. This is a canonicalization that is more involved than
-// necessary across the boundary of regions without captures.
-static void pullInProducersInSameGroup(
-    PatternRewriter &rewriter, IREE::Flow::DispatchWorkgroupsOp dispatchOp,
-    linalg::LinalgOp rootOp, int64_t groupNum) {
-  LLVM_DEBUG(llvm::dbgs() << "pull in producers for op: " << rootOp << "\n");
-
-  // Scoped within DispatchWorkgroupOp.
-  OpBuilder::InsertionGuard g(rewriter);
-  rewriter.setInsertionPointToStart(&dispatchOp.getRegion().front());
-  for (auto en : llvm::enumerate(rootOp->getOperands())) {
-    if (auto producer = en.value().getDefiningOp<linalg::LinalgOp>()) {
-      if (!isInFusionGroup(producer, groupNum)) continue;
-      DEBUG_WITH_TYPE(DEBUG_TYPE,
-                      llvm::dbgs() << "current producer: " << producer << "\n");
-
-      Operation *fusedProducer = rewriter.clone(*producer);
-      rewriter.replaceOpWithinBlock(producer, fusedProducer->getResults(),
-                                    &dispatchOp.getRegion().front());
-      removeFusionGroupsAttribute(fusedProducer);
-
-      pullInProducersInSameGroup(rewriter, dispatchOp, fusedProducer, groupNum);
-    } else if (auto producer = en.value().getDefiningOp<tensor::PadOp>()) {
-      DEBUG_WITH_TYPE(DEBUG_TYPE,
-                      llvm::dbgs() << "current producer: " << producer << "\n");
-
-      Operation *fusedProducer = rewriter.clone(*producer);
-      rewriter.replaceOpWithinBlock(producer, fusedProducer->getResults(),
-                                    &dispatchOp.getRegion().front());
+/// Returns the list of operations that are to be cloned into the dispatch
+/// based on the root operation.
+static SmallVector<Operation *> getOperationsToMoveIntoDispatch(
+    Operation *rootOp) {
+  SmallVector<Operation *> dispatchOps;
+  dispatchOps.push_back(rootOp);
+  if (!hasRootOpAttribute(rootOp)) return dispatchOps;
+
+  int64_t groupNum = getRootNumber(rootOp);
+  std::deque<Operation *> worklist;
+  worklist.push_back(rootOp);
+  llvm::SmallDenseSet<Operation *, 2> movedOps;
+  movedOps.insert(rootOp);
+
+  while (!worklist.empty()) {
+    Operation *currRoot = worklist.front();
+    worklist.pop_back();
```

Comment:
`pop_back()` or `pop_front()`?

Reply:
Good catch. Thats why I used a `deque` in the first place! :D